mono/mini/mini-runtime.c

```diff
@@ -1905,6 +1905,97 @@ mono_jit_map_is_enabled (void)
 
 #endif
 
+#if ENABLE_JIT_DUMP
+static FILE* perf_dump_file;
+uint64_t codeIndex = 0;
+pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+void *mmapAddr = MAP_FAILED;
+
+void
+mono_enable_jit_dump (void)
+{
+        if (!perf_dump_file) {
+		char name [64];
+		FileHeader * header = malloc (sizeof (FileHeader));
+		int result = 0;
+		
+		result = pthread_mutex_lock (&mutex);
+		
+		g_snprintf (name, sizeof (name), "/tmp/jit-%d.dump", getpid ());
+		unlink (name);
+		perf_dump_file = fopen (name, "w");
+		
+		add_file_header_info (header);
+		fwrite (header, sizeof (FileHeader), 1, perf_dump_file);
+		
+		mmapAddr = mmap(NULL, sizeof (FileHeader), PROT_READ | PROT_EXEC, MAP_PRIVATE, 1, 0);
+		
+		result = pthread_mutex_unlock (&mutex);
+	}
+}
+
+void
+add_file_header_info (FileHeader *header)
+{
+	header->magic = JIT_DUMP_MAGIC;
+	header->version = JIT_DUMP_VERSION;
+	header->total_size = sizeof (FileHeader);
+	header->elf_mach = ELF_MACHINE;
+	header->pad1 = 0;
+	header->pid = getpid ();
+	header->timestamp = getTimeStampNS ();
+	header->flags = 0;
+}
+
+guint64
+getTimeStampNS (void)
+{
+	struct timespec ts;
+	int result = clock_gettime (CLOCK_MONOTONIC, &ts);
+	return  ts.tv_sec * 1000000000ULL + ts.tv_nsec;
+}
+
+void
+mono_emit_jit_dump (MonoJitInfo *jinfo, gpointer code)
+{
+	int result = 0;
+	
+	if (perf_dump_file) {
+		JitCodeLoadRecord *record = malloc (sizeof (JitCodeLoadRecord));
```

Comment:
`record` leaks. Please allocate it on the stack (same idea for the header).

Reply:
Please make it value C++, such as with `g_new` or add a cast.

Reply:
No need to use `g_new()` or add a cast; just define it in the stack and initialize it:

```c
JitCodeLoadRecord record;

record.foo = ...;
record.bar = ...;
```

(No need to free this afterwards.)